### PR TITLE
Fix PaneFocused notification storm on pane destruction and mux reconciliation

### LINKED ISSUE #4390: Switching panes in multiple directions at once causes the focused pane to rapidly switch when any key is pressed
### What Operating System(s) are you seeing this problem on?

Linux Wayland

### Which Wayland compositor or X11 Window manager(s) are you using?

KDE Plasma 5.27.8

### WezTerm version

20231002-075348-a61cc8eb

### Did you try the latest nightly build to see if the issue is better (or worse!) than your current version?

Yes, and I updated the version box above to show the version of the nightly that I tried

### Describe the bug

If I try to change to two different panes e.g. `Alt+Left+Right` (using my keybinds), the pane switching completely breaks. It doesn't always happen, but it happens often enough to be a huge pain.

It's easiest to explain with a video:

https://github.com/wez/wezterm/assets/114271894/99963be2-2517-4779-a976-938c12504d46

First I spam switch panes with `Alt+Random Arrow Key` (custom keybind), then once it breaks I spam a random key (z). You can see the focus randomly flicker between panes.

### To Reproduce

1. Use my configuration
2. Create multiple panes (`Ctrl+p -> s -> <Arrow Key>` with my config, but I don't think it really matters how they're made)
3. Switch to multiple panes at once using `Alt+<Up>+<Right>` or some other combination of arrow keys. There must a pane in the directions pressed for it to break.

### Configuration

local wezterm = require 'wezterm'
local act = wezterm.action

return {

  -- Keyboard Shortcuts
  leader = { key = 'p', mods = 'CTRL', timeout_milliseconds = 5000 },

  keys = {
    -- Key Tables
    { key = 'r', mods = 'LEADER', action = act.ActivateKeyTable { name = 'resize_pane', one_shot = false } },
    { key = 's', mods = 'LEADER', action = act.ActivateKeyTable { name = 'split_pane', one_shot = true } },
    { key = 'm', mods = 'LEADER', action = act.ActivateKeyTable { name = 'move_pane', one_shot = false } },

    -- Close Active Pane
    { key = 'w', mods = 'LEADER', action = act.CloseCurrentPane { confirm = false } },    
    -- Change Active Pane
    { key = 'LeftArrow', mods = 'ALT', action = act.ActivatePaneDirection 'Left' },
    { key = 'RightArrow', mods = 'ALT', action = act.ActivatePaneDirection 'Right' },
    { key = 'UpArrow', mods = 'ALT', action = act.ActivatePaneDirection 'Up' },
    { key = 'DownArrow', mods = 'ALT', action = act.ActivatePaneDirection 'Down' },

    
  },

  key_tables = {
    split_pane = {
      { key = 'LeftArrow', action = act.SplitPane { direction = 'Left' } },
      { key = 'RightArrow', action = act.SplitPane { direction = 'Right' } },
      { key = 'r', action = act.SplitPane { direction = 'Right' } },
      { key = 'DownArrow', action = act.SplitPane { direction = 'Down' } },
      { key = 'd', action = act.SplitPane { direction = 'Down' } },
      { key = 'UpArrow', action = act.SplitPane { direction = 'Up' } },

      { key = 'Escape', action = 'PopKeyTable' },
    },
  }
}

### Expected Behavior

Pane switches without issue

### Logs

14:30:54.988  ERROR  wezterm_mux_server_impl::local > writing pdu data buffer: Broken pipe (os error 32)
12:43:27.272  ERROR  wezterm_gui::frontend > Error reconciling PaneFocused notification: pane 0 not found
12:43:27.272  ERROR  wezterm_gui::frontend > Error reconciling PaneFocused notification: pane 1 not found

### Anything else?

* The `broken pipe` error from the logs doesn't always appear. Additionally, if too many inputs are given in a short time, the focus will keep flickering after I stop pressing keys for a bit. The message doesn't seem to appear until _after_ the flickering stops. It always appears once.
* The `pane not found` errors in the logs only appear once I close the wezterm GUI window.
* Zooming in then out of a pane with `Ctrl+Shift+z` fixes the main problem, but if a TUI (helix, in my case) is open, the pane with the TUI sometimes (not always) become unresponsive and needs to be closed anyways. 
* If I use one keybind at a time (e.g. `Alt+Right` then `Alt+Up`) instead of combining them (`Alt+Right+Up`) then it doesn't break.
* I run into this issue _far_ more frequently on my laptop than on my desktop despite using the keybinds the same amount, to the point where I'm willing to use panes despite the risk on my desktop, but avoid them on my laptop. I assume it's due to the hardware difference (`i7-1185G7 + 32GB RAM` vs `7950x3d + 64GB RAM + 6900xt`), but I'm not sure.

### PR DESCRIPTION
**Pull request by fculpo**
Fixes #4390

Picks up where #4737 left off. That PR was closed because
unconditionally suppressing the `PaneFocused` notification from
`advise_focus_change` broke `wezterm cli activate-{tab,pane}` over SSH
and prevented focus updates from reaching other attached clients.

## Problem

Three things conspire to create a runaway focus-notification loop:

1. **Destroyed-pane storm** — when a unix-domain detach tears down many
   panes at once, the subscriber drains a burst of `PaneFocused`
   notifications whose pane ids are already gone. The handler in
   `frontend.rs` spawns a reconciliation task on the main thread for
   *every* one of them. Thousands of these pile up faster than the UI
   thread can service them, resulting in unresponsive mouse input,
   erratic focus changes, and a flood of
   `"Error reconciling PaneFocused notification: pane X not found"`
   in the log.

2. **Redundant re-notification** — several call sites
   (`Mux::add_pane_to_window`, `sessionhandler`, `tmux_commands`)
   call `set_active_pane` and then immediately fire their own
   `MuxNotification::PaneFocused`. Because `advise_focus_change` also
   fires the same notification, every focus change emits the event
   twice. With two rapid user-initiated switches (pane 0 → 1 → 0),
   the doubled events interleave and the queue never drains —
   exactly the scenario described in #4390 and #4517.

3. **Client echo-back** — when a remote `PaneFocused` arrives at a
   client pane, `focus_pane_and_containing_tab` calls
   `focus_changed(true)`, which for a `ClientPane` advises the server
   of the "new" focus. This echoes the server's own notification
   right back as a fresh `SetFocusedPane` PDU, adding fuel to the
   loop.

## Fix

Each layer gets a targeted guard rather than a blanket suppression:

- **`frontend.rs`** — cheap synchronous `get_pane()` check before
  spawning the reconciliation task. Defunct notifications are dropped
  without touching the UI thread. A pane that disappears between the
  guard and the spawned task falls through to the existing error log.

- **`tab.rs`** — new `NotifyMux` enum and `set_active_pane_with_notify`
  entry point. Call sites that fire their own notification pass
  `NotifyMux::No` to avoid the duplicate; everyone else (including the
  GUI and `set_active_idx`) still gets the notification via the
  default `set_active_pane` → `NotifyMux::Yes` path.
  This is the same `NotifyMux` idea from #4737, but applied only to the
  three call sites that already send the notification themselves, so
  SSH-attached clients and `wezterm cli activate-*` keep working.

- **`clientpane.rs`** — pre-set `focused_remote_pane_id` before calling
  `focus_pane_and_containing_tab`, so the subsequent
  `focus_changed(true)` sees the pane as already focused on the server
  side and does not echo back a `SetFocusedPane` PDU.

## Testing

Added `set_active_pane_can_suppress_mux_notification` unit test that
verifies `NotifyMux::No` actually suppresses the notification.

Manually verified:
- local pane focus switching (including rapid Alt+arrow spam)
- `wezterm cli activate-pane` / `activate-tab` over unix domain
- detaching a domain with many panes no longer floods the log

### CODE REVIEW

File: mux/src/tab.rs [hunk 1]
```diff
@@ -2517,6 +2531,53 @@ mod test {
...
+        };
+
+        let mux = Arc::new(Mux::new(None));
+        Mux::set_mux(&mux);
+
+        let notifications = Arc::new(Mutex::new(Vec::new()));
+        let seen = Arc::clone(&notifications);
+        mux.subscribe(move |notification| {
+            if let MuxNotification::PaneFocused(pane_id) = notification {
+                seen.lock().push(pane_id);
```
Comment (bew) [inline mux/src/tab.rs:2551]: You set `seen` but never attempt to read it later on, is that intentional? 🤔

File: wezterm-gui/src/frontend.rs [hunk 2]
```diff
@@ -78,13 +78,20 @@ impl GuiFrontEnd {
...
-                            log::error!("Error reconciling PaneFocused notification: {err:#}");
-                        }
-                    })
-                    .detach();
+                    // Skip notifications for panes the mux has already destroyed
+                    // (e.g. backlog draining after a domain detach). Prevents the
+                    // UI thread from being flooded with no-op reconciliation tasks
+                    // when many panes go away at once.
+                    // See https://github.com/wezterm/wezterm/issues/4390
+                    if Mux::get().get_pane(pane_id).is_some() {
```
Comment (bew) [inline wezterm-gui/src/frontend.rs:86]: this impl might be right, but I'm not sure why some users kept getting `Error reconciling PaneFocused notification: pane X not found` in https://github.com/wezterm/wezterm/issues/4390, they weren't destroying panes AFAIK 🤔

do you have any idea on the reason?

File: wezterm-client/src/pane/clientpane.rs [hunk 3]
```diff
@@ -226,6 +226,17 @@ impl ClientPane {
                 // it here.
                 log::trace!("advised of remote pane focus: {pane_id}");
 
+                // Match the server's focus state before applying it locally.
+                // `focus_pane_and_containing_tab` calls through to
+                // `focus_changed(true)`, and for remote panes that normally
+                // advises the server of the new focus. Without this guard, a
+                // server-originated focus notification can be echoed back as a
+                // fresh `SetFocusedPane` request.
```
Comment (bew) [inline wezterm-client/src/pane/clientpane.rs:234]: please link to the issue for reference 🙏

File: mux/src/tab.rs [hunk 4]
```diff
@@ -2517,6 +2531,53 @@ mod test {
         assert_eq!(600, panes[2].pixel_height);
     }
 
+    #[test]
+    fn set_active_pane_can_suppress_mux_notification() {
```
Comment (bew) [inline mux/src/tab.rs:2535]: could you add some comments in this test about the different phases and what should / should not happen regarding mux notifs? 🙏

File: mux/src/lib.rs [hunk 5]
```diff
@@ -559,7 +559,7 @@ impl Mux {
             .get_tab(tab_id)
             .ok_or_else(|| anyhow::anyhow!("tab {tab_id} not found"))?;
 
-        tab.set_active_pane(&pane);
+        tab.set_active_pane_with_notify(&pane, NotifyMux::No);
```
Comment (bew) [inline mux/src/lib.rs:562]: please add a small comment explaining why we don't notify mux here 🙏

File: mux/src/tab.rs [hunk 6]
```diff
@@ -1768,22 +1778,26 @@ impl TabInner {
         {
             self.active = item.index;
             self.recency.tag(item.index);
-            self.advise_focus_change(prior);
+            self.advise_focus_change(prior, notify);
         }
     }
 
-    fn advise_focus_change(&mut self, prior: Option<Arc<dyn Pane>>) {
+    fn advise_focus_change(&mut self, prior: Option<Arc<dyn Pane>>, notify: NotifyMux) {
```
Comment (bew) [inline mux/src/tab.rs:1785]: could you rename `notify` to `notify_mux` in all functions, so it's extra explicit while reading what's being given / checked 🙏

File: wezterm-mux-server-impl/src/sessionhandler.rs [hunk 7]
```diff
@@ -358,7 +358,7 @@ impl SessionHandler {
                             let tab = mux
                                 .get_tab(tab_id)
                                 .ok_or_else(|| anyhow::anyhow!("tab {tab_id} not found"))?;
-                            tab.set_active_pane(&pane);
+                            tab.set_active_pane_with_notify(&pane, NotifyMux::No);
```
Comment (bew) [inline wezterm-mux-server-impl/src/sessionhandler.rs:361]: similarly, add a small comment mentioning why we don't notify mux here 🙏

File: mux/src/tmux_commands.rs [hunk 8]
```diff
@@ -341,7 +341,7 @@ impl TmuxDomainState {
 
                     match mux.get_tab(local_tab.tab_id) {
                         Some(tab) => {
-                            tab.set_active_pane(&local_pane);
+                            tab.set_active_pane_with_notify(&local_pane, NotifyMux::No);
```
Comment (bew) [inline mux/src/tmux_commands.rs:344]: and here 🙏

File: Cargo.toml [hunk 9]
```diff
@@ -58,7 +58,7 @@ camino = "1.0"
 cassowary = "0.3"
 cc = {version="1.0", features = ["parallel"]}
 cgl = "0.3"
-chrono = {version="0.4", default-features=false, features=["unstable-locales", "serde"]}
+chrono = {version="0.4", default-features=false, features=["unstable-locales", "serde", "clock"]}
```
Comment (bew) [inline Cargo.toml:61]: what is this change for? 🤔

File: mux/src/tab.rs [hunk 10]
```diff
@@ -699,7 +705,11 @@ impl Tab {
     }
 
     pub fn set_active_pane(&self, pane: &Arc<dyn Pane>) {
```
Comment (bew) [inline mux/src/tab.rs:707]: what do you think about completely removing `set_active_pane` (or add `notify_mux` to that function) and always force caller to think about mux notification or not (and document when not) 🤔

**Review by bew — COMMENTED** [review]
A few comments & questions 🤔

~~Note that WezTerm is currently on hold/unmaintained, Wez hasn't been responding to PRs for a while. 
👉 I'm interested in starting a community fork, see https://github.com/wezterm/wezterm/issues/6341#issuecomment-4483047032, so this might gets merged there once ready.~~

EDIT: FYI Development/Maintenance on Wezterm is slowly coming back on track! ([ref](https://github.com/wezterm/wezterm/issues/6341#issuecomment-4620396384))

**Comment by Fail-Safe** [comment]
Confirmed this fixes my reproducible pane-focus hang on macOS.

Before the patch, with pane_focus_follows_mouse = true and any registered format-tab-title callback—even one returning only "x"—rapidly moving the mouse between panes on the same tab reliably caused WezTerm to lock up at 100% CPU.

I cherry-picked this PR onto current upstream main, ran the added mux test, and built a release version. After applying the patch, the reproducer no longer hangs. I also restored my full tab-title formatter and repeated the stress test successfully.

Thanks for working on this!

**Comment by bew** [comment]
Hello @fculpo :)

Thanks again for the PR, are you willing to get it to the finish line and get it merged, and start answering comments from my [last review](https://github.com/wezterm/wezterm/pull/7763#pullrequestreview-4333659248) ? 🤔

Or maybe @ankitson since you made https://github.com/ankitson/wezterm/commit/f3c52f3ba189b6e4c52fd28c02e081f5c1c353b8 that addresses most of my comments above, maybe you'd want to open a separate PR? 🤔

**Comment by ankitson** [comment]
Hi @bew :)

Yes there are a couple of bugs in Wezterm that annoyed me too much so I had to fix them.. i just wanted to review them a bit better before sending them to you - will open PRs for them soon!

**Comment by ankitson** [comment]
@bew My PR is ready on #7871
